jupyterhub_traefik_proxy/fileprovider.py

```diff
@@ -66,16 +66,36 @@ def _load_dynamic_config(self):
         except FileNotFoundError:
             dynamic_config = {}
 
-        if not dynamic_config:
-            dynamic_config = {
-                "http": {"services": {}, "routers": {}},
-                "jupyter": {"routers": {}},
-            }
+        # fill in default keys
+        # use setdefault to ensure these are always fully defined
+        # and never _partially_ defined
+        http = dynamic_config.setdefault("http", {})
+        http.setdefault("services", {})
+        http.setdefault("routers", {})
+        jupyter = dynamic_config.setdefault("jupyter", {})
+        jupyter.setdefault("routers", {})
         return dynamic_config
 
     def persist_dynamic_config(self):
         """Save the dynamic config file with the current dynamic_config"""
-        self.dynamic_config_handler.atomic_dump(self.dynamic_config)
+        # avoid writing empty dicts, which traefik doesn't handle for some reason
+        dynamic_config = self.dynamic_config
+        if (
+            not dynamic_config["http"]["routers"]
+            or not dynamic_config["http"]["services"]
+        ):
+            # traefik can't handle empty dicts, so don't persist them.
+            # But don't _remove_ them from our own config
+            # I think this is a bug in traefik - empty dicts satisfy the spec
+            # use shallow copy, which is cheap but most be done at every level where we modify keys
+            dynamic_config = dynamic_config.copy()
+            dynamic_config["http"] = http = dynamic_config["http"].copy()
+            for key in ("routers", "services"):
+                if not http[key]:
+                    http.pop(key)
+            if not http:
+                dynamic_config.pop("http")
```

Comment:
Shou;dn't we handle the same way `dynamic_config["jupyter"]` too?

Reply:
I don't believe so, because it's traefik, not the underlying file store, that doesn't handle it right. And traefik doesn't read the jupyter key

Reply:
Right, the `jupyter` key was just for us to store the additional `data`